src/main/scala/com/fulcrumgenomics/bam/SamRecordClipper.scala

```diff
@@ -0,0 +1,303 @@
+import scala.collection.mutable.ListBuffer
+
+/**
+  * Provides a suite of methods for clipping (soft- and hard-) bases from the beginnings
+  * and ends of reads in various ways.  Cigar strings, bases, qualities and alignment
+  * positions are all correctly adjusted post-clipping.
+  *
+  * Note that there are several "flavours" of method:
+  *   - Ones that work on the start and end of the read in whatever orientation the read
+  *     is in, vs. working on the 5' or 3' end of the read
+  *   - Ones that attempt to clip an additional N bases beyond any clipping already
+  *     provided (clip[Start|End|5PrimeEnd|3PrimeEnd]OfAlignment) and ones that
+  *     attempt to make it so that the read has N bases clipped, including any existing
+  *     clipping (clip[Start|End|5PrimeEnd|3PrimeEnd]OfRead)
+  */
+object SamRecordClipper {
+  /** An enumeration representing the various ways to clip bases within a read. */
+  object ClippingMode extends Enumeration { val Soft, Hard, SoftWithMask = Value }
+  type ClippingMode = ClippingMode.Value
+
+  private val NoCallBase = 'N'.toByte
+  private val NoCallQual = 2.toByte
+
+  /**
+    * Adds clipping of _at least_ numberOfBasesToClip to the start (left hand end)
+    * of an alignment. If clipping already exists at the start of the read it is
+    * preserved, and numberOfBasesToClip more clipping is added.
+    *
+    * If is unmapped or the numberOfBasesToClip is < 1 nothing is done.
+    *
+    * If the read has fewer clippable bases than requested clipping, the read is
+    * unmapped.
+    *
+    * If hard-clipping is requested and the read has existing soft-clipping at the start
+    * it is converted to hard-clipping before adding further clipping.
+    *
+    * If soft-clipping-with-masking is requested and the read already contains soft
+    * clipping at the start of the read, both the existing and new soft clipped bases
+    * are masked.
+    *
+    * @param rec the record to be clipped
+    * @param numberOfBasesToClip the number of additional bases to clip beyond any clipping
+    *                            already applied to the read
+    * @param mode the kind of clipping to be performed
+    */
+  def clipStartOfAlignment(rec: SAMRecord, numberOfBasesToClip: Int, mode: ClippingMode) : Unit = {
+    if (rec.getReadUnmappedFlag || numberOfBasesToClip < 1) {
+      Unit
+    }
+    else if (numberOfClippableBases(rec) <= numberOfBasesToClip) {
+      SAMUtils.makeReadUnmapped(rec)
+    }
+    else {
+      val oldElems = rec.getCigar.toSeq
+      val oldBases = rec.getReadBases
+      val oldQuals = rec.getBaseQualities
+      val (newElems, readBasesClipped, refBasesClipped, bases, quals) = clip(oldElems, numberOfBasesToClip, oldBases, oldQuals, mode)
+      rec.setAlignmentStart(rec.getAlignmentStart + refBasesClipped)
+      rec.setCigar(new Cigar(util.Arrays.asList(newElems:_*)))
+      rec.setReadBases(bases)
+      rec.setBaseQualities(quals)
```

Comment:
I think I'd rather leave that outside of this code - e.g. TrimPrimers should probably do that since it's writing a BAM file and you don't want an invalid BAM.